### PR TITLE
HWKAPM-751 Provide a URI templating capability to convert path and qu…

### PR DESCRIPTION
**Pull request by objectiser**
…ery parameters into node properties, and use a stable URI value for aggregation purposes

### CODE REVIEW

**Comment by objectiser** [comment]
@jpkrohling Could you review please?

As mentioned in the jira, I think it would be good to have a mechanism in the OT provider that can handle dealing with URI templates - and then as a separate jira we could look at obtaining those templates from jax-rs annotations etc.

This property driven approach will enable other mechanisms, including the application or custom agent rules, to also be used to define the template on a span created by other instrumentation techniques.

File: api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java [hunk 1]
```diff
@@ -73,6 +89,119 @@ public static boolean isOriginalURI(Node node, String uri) {
...
+        // If query parameter template defined, then process
+        if (!queryParameters.isEmpty() && node.hasProperty(Constants.PROP_HTTP_QUERY)) {
+            StringTokenizer st = new StringTokenizer(node.getProperties(Constants.PROP_HTTP_QUERY).iterator().next().getValue(), "&");
+            while (st.hasMoreTokens()) {
+                String token = st.nextToken();
+                String[] namevalue = token.split("=");
+                if (namevalue.length == 2) {
+                    if (queryParameters.contains(namevalue[0])) {
+                        try {
+                            node.getProperties().add(new Property(namevalue[0],
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java]: Shouldn't you decode the `namevalue[0]` as well?

File: api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java [hunk 2]
```diff
@@ -73,6 +89,119 @@ public static boolean isOriginalURI(Node node, String uri) {
...
+                    } else {
+                        log.severe("Expecting query parameter template, e.g. {name}, but got '" + token + "'");
+                    }
+                }
+            }
+
+            StringTokenizer uriTokens = new StringTokenizer(node.getUri(), "/");
+            StringTokenizer templateTokens = new StringTokenizer(template, "/");
+
+            if (uriTokens.countTokens() == templateTokens.countTokens()) {
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java]: This means that you are ruling out parameters that contain slashes themselves. For instance, I think this is possible in JAX-RS:

URL: `/download/2017/01/18/scanned-file.pdf`
Template: `/download/{docPath}`
`docPath` then gets translated as `2017/01/18/scanned-file.pdf`

Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java]: Is this possible in jax-rs? Either way, would it be ok if I raise a separate jira to check/implement, as I would like to get this in asap for some demos.

Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java]: I think so. A quick check on the docs didn't yield any positive results, but I remember seeing this somewhere. I can check it later, but I think it's fine to merge it without this use case.

**Comment by objectiser** [comment]
@jpkrohling PR updated

**Comment by jpkrohling** [comment]
LGTM

**Comment by objectiser** [comment]
@jpkrohling Thanks!

File: api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java (hunk 2 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java]: If you find it is possible, then raise a jira so we can add support.

Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/utils/NodeUtil.java]: https://issues.jboss.org/browse/HWKAPM-851

**Pull request merged**